### DISCUSSION TITLE
Submit single input field (not form) via AJAX?

### DISCUSSION THREAD

**Discussion by tedyoung** [unpoly/unpoly#201] [Q&A]
I have a field on a form that is used to pre-fill two other input fields, e.g., ZIP code filling in the State and City. However, there's only the one Form, so I don't want a whole form submitted. The "observe" functionality seems close, but I don't want the content sent to the server until the user hits a button (vs. as they type).

**Comment by adam12** [comment]
How about [`up-validate`](https://unpoly.com/input-up-validate)? It will submit the entire form but you don't need to refresh the entire form.

**Reply by tedyoung** [reply]
I looked at up-validate, but it submits the whole form to the form's endpoint and _expects a new form back_. This isn't really possible to do in my stack (Spring MVC with Thymeleaf), at least not without a lot of work that would just be easier to do in raw JavaScript (how I currently do it).

I understand this may be beyond Unpoly's scope.

Also, the docs are unclear as to _when_ the form gets sent:

> When a form field with this attribute is changed

I _assume_ that happens upon loss of focus for that field? Or does that happen as a pre-flight-like request when the form's submit button is clicked?

**Reply by adam12** [reply]
How about `up-observe` paired with `up.render` (presuming you're on v2)?

Untested, but something like:

```html
up-observe="up.render({ target: 'input#state, select#country', url: '/the-url', params: { zip: value }})"
```

EDIT: I forgot `up-observe` fires continuously, so I'm not sure this would work for you as you want.

**Reply by triskweline** [reply]
> I looked at up-validate, but it submits the whole form to the form's endpoint and expects a new form back. This isn't really possible to do in my stack (Spring MVC with Thymeleaf), at least not without a lot of work that would just be easier to do in raw JavaScript (how I currently do it).

Unfortunately I have no experience with Spring or Thymeleaf. The idea behind `[up-validate]`'s behavior was that classic server-side apps already has endpoints that take a form submission and re-render the form with errors marks if any input is invalid. We wanted to re-use this endpoint for the purpose of 

> Untested, but something like:
> ```
> up-observe="up.render({ target: 'input#state, select#country', url: '/the-url', params: { zip: value }})"
> ```
> EDIT: I forgot up-observe fires continuously, so I'm not sure this would work for you as you want.

If your expression returns a promise, Unpoly will wait with additional changes until that promise settles:

```html
<input up-observe="return up.render(....)">
```

If more than one render calls are delayed in that way, all but the latest will be discarded.
This way only a single validation request is in flight at a given time.

**Reply by tedyoung** [reply]
Thanks for the suggestions. I've only been toying with Unpoly to see if it's a fit for server-side generated HTML using Java+Spring MVC+Thymeleaf (vs. using Vue or similar). The proposed ideas seem unwieldy enough for me to stick with my plain vanilla solution.

> The idea behind [up-validate]'s behavior was that classic server-side apps already has endpoints that take a form submission and re-render the form with errors marks if any input is invalid.

I get that, but the Java-based frameworks that I've used re-render the entire page and return it whole; they can't return just the form fragment (well, you could force it, but then it's non-standard).

**Reply by adam12** [reply]
> I get that, but the Java-based frameworks that I've used re-render the entire page and return it whole; they can't return just the form fragment (well, you could force it, but then it's non-standard).

Just FYI, but rendering _only_ the form fragment isn't required. It can be in a full page, as if it was rendered normally.

> stick with my plain vanilla solution.

Out of curiosity, can you share what you have?

**Comment by triskweline** [comment]
> The proposed ideas seem unwieldy enough for me to stick with my plain vanilla solution.

I'm not sure if you understood how form validations work in Unpoly.

[Here is a tutorial for Spring MVC and Thymeleaf](https://www.wimdeblauwe.com/blog/2021/05/23/form-handling-with-thymeleaf/)

Scroll down to *"To recap, this is the full source code of the UserController"* and you will see the following method that saves a user form:

```java
@Controller
@RequestMapping("/users")
public class UserController {
    ...

    @PostMapping("/create")
    public String doCreateUser(@Valid @ModelAttribute("formData") CreateUserFormData formData,
                               BindingResult bindingResult,
                               Model model) {
        if (bindingResult.hasErrors()) {
            return "users/create";
        }

        service.createUser(formData.toParameters());

        return "redirect:/users";
    }

    ...
}
```

To use Unpoly's live validations we only need to check if the [`X-Up-Validate`](https://unpoly.com/X-Up-Validate) request header contains a value. If yes, we validate and re-render the form:

```java
@Controller
@RequestMapping("/users")
public class UserController {
    ...

    @PostMapping("/create")
    public String doCreateUser(@Valid @ModelAttribute("formData") CreateUserFormData formData,
                               BindingResult bindingResult,
                               Model model,
                               @RequestHeader(value="X-Up-Validate", required=false) String validatingField) {
        if (bindingResult.hasErrors() || validatingField != null) {
            return "users/create";
        }

        service.createUser(formData.toParameters());

        return "redirect:/users";
    }

    ...
}
```

This is ~100 characters for live form validation that re-uses most the code you already have.
This does not look unwieldy to me.
